### PR TITLE
feat: support elasticsearch vector database

### PR DESCRIPTION
**Pull request by miendinh**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Add Elasticsearch vector database for RAG, this feature requires package `elasticsearch==7.15.2`

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by miendinh** [comment]
hi @JohnJyong
do you have any comments?

**Comment by crazywoola** [comment]
Hi there, please run `./dev/reformat` in root dir. The CI build didn't pass.

**Comment by miendinh** [comment]
Could you add `Setup ElasticSearch' to the CI like Milvus or Quant? @crazywoola

**Comment by JohnJyong** [comment]
please fix the pytest and add vdb integration test for it ， thanks @miendinh 
intergration test folder ： ‘tests.integration_tests.vdb’

**Comment by miendinh** [comment]
> please fix the pytest and add vdb integration test for it ， thanks @miendinh intergration test folder ： ‘tests.integration_tests.vdb’

added, but need ElasticSearch on CI to run ES tests, plz check @JohnJyong

**Comment by JohnJyong** [comment]
CI workflow is set up in the .github/workflows/api-tests.yml @miendinh

**Comment by miendinh** [comment]
Hi @JohnJyong, I implemented the integration test and ran it on my local machine successfully:

image

However, the Pytest on the CI, it always returns:

`FAILED api/tests/integration_tests/vdb/elasticsearch/test_elasticsearch.py::test_elasticsearch_vector - elastic_transport.ConnectionError: Connection error caused by: ConnectionError(Connection error caused by: ProtocolError(('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))))
`
I assume that there are no firewalls dropping the requests to port 9200 of the ES container. Please help me check!
Thanks in advance. :-)

**Comment by QIN2DIM** [comment]
It might be better to add `kibana` to the same docker-comopse config file 🤔

**Comment by JohnJyong** [comment]
you  does't setup the `ELASTICSEARCH_API_KEY_ID ` and `ELASTICSEARCH_API_KEY` in your es docker-compose file @miendinh

**Comment by taohongxiu** [comment]
es version 8.12.1

**Comment by bowenliang123** [comment]
How about rebase the changes to the main branch first? And let's see what to do in follow-ups, like bumping the ES version to v8 and check the connection exception in details.

**Comment by crazywoola** [comment]
Hello please fix the errors in CI and rebase the code from main branch.

**Comment by miendinh** [comment]
> How about rebasing the changes to the main branch first? And let's see what to do in follow-ups, like bumping the ES version to v8 and checking the connection exception in detail.

@bowenliang123 rebased, plz check :-)

**Comment by takatost** [comment]
Sorry for the late reply, but still facing issues with test failures and conflicts. Can you please fix them?
P.S. The current version has both pip and poetry installed. Can you also synchronize the dependencies in requirements.txt and pyproject.toml?

**Comment by miendinh** [comment]
> Sorry for the late reply, but still facing issues with test failures and conflicts. Can you please fix them? P.S. The current version has both pip and poetry installed. Can you also synchronize the dependencies in requirements.txt and pyproject.toml?

fixed, dify CI does not make sense, I always get this error:

`FAILED api/tests/integration_tests/vdb/es/test_elasticsearch.py::test_elasticsearch_vector - elastic_transport.ConnectionError: Connection error caused by: ConnectionError(Connection error caused by: ProtocolError(('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))))`

but in my local dev,  test_elasticsearch_vector is fine.

image

**Comment by bowenliang123** [comment]
My suggestion would be test it locally first with the exact middleware docker compose file. Furthermore, make sure  all the versions of client and sever side(mixed use of es7 and es8), and authentication details.

**Comment by miendinh** [comment]
finally, It works.

plz check @crazywoola

**Review by crazywoola — DISMISSED** [review]
LGTM

**Review by crazywoola — DISMISSED** [review]

File: api/pyproject.toml [hunk 1]
```diff
@@ -176,6 +176,7 @@ werkzeug = "~3.0.1"
 xinference-client = "0.9.4"
 yarl = "~1.9.4"
 zhipuai = "1.0.7"
+elasticsearch = "8.12.1"
```
Comment (bowenliang123) [inline api/pyproject.toml]: Is it using ElasticSearch v8 client to accessing v7 server in the test? Please avoid crossing versions for validation.

File: api/poetry.lock
Comment (bowenliang123) [inline api/poetry.lock:1]: Please keep unrelated  packages' pinned version untouched. Revert the changes and then use dev/sync-poetry script to align the necessary changes in poetry lockfile.

**Review by bowenliang123 — CHANGES_REQUESTED** [review]
Please check again.

**Comment by miendinh** [comment]
plz check @crazywoola , I will improve for supporting es 8.x later :-), thanks

**Review by bowenliang123 — CHANGES_REQUESTED** [review]
The configs for ES is never declared in DifyConfig class. It makes them are never parsed properly from env file and not accessisble in runtime.

And I strongly encourage you to implement the ES vdb feature on the baseline of v8 rather than v7, for both the SDK client version and the server version. ES v8 has breaking changes and packs for enhancement in vector search, which is more modern and production ready for practical use.

File: dev/pytest/pytest_vdb.sh [hunk 2]
```diff
@@ -8,4 +8,5 @@ pytest api/tests/integration_tests/vdb/chroma \
   api/tests/integration_tests/vdb/pgvector \
   api/tests/integration_tests/vdb/qdrant \
   api/tests/integration_tests/vdb/weaviate \
+  api/tests/integration_tests/vdb/es \
```
Comment (bowenliang123) [inline dev/pytest/pytest_vdb.sh]:
```suggestion
  api/tests/integration_tests/vdb/elasticsearch \
```

Use the full name for path is more clear and extensible .

File: docker/docker-compose.yaml [hunk 3]
```diff
@@ -553,6 +553,26 @@ services:
...
+      - "9200:9200"
+  kibana:
+    container_name: kb-container
+    image: docker.elastic.co/kibana/kibana:7.17.21
+    environment:
+      - ELASTICSEARCH_HOSTS=http://elasticsearch:9200
+    depends_on:
+      - elasticsearch
+    ports:
+      - "5601:5601"
```
Comment (bowenliang123) [inline docker/docker-compose.yaml]: Please remove it , as Kibana is not required for running the Elasticsearch server.

File: api/core/rag/datasource/vdb/es/elasticsearch_vector.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+    def search_by_vector(self, query_vector: list[float], **kwargs: Any) -> list[Document]:
+        query_str = {
+            "query": {
+                "script_score": {
+                    "query": {
+                        "match_all": {}
+                    },
+                    "script": {
+                        "source": "cosineSimilarity(params.query_vector, 'vector') + 1.0",
```
Comment (bowenliang123) [inline api/core/rag/datasource/vdb/es/elasticsearch_vector.py]: What's the meaning here by seeting `1.0` ?

Comment (miendinh) [reply api/core/rag/datasource/vdb/es/elasticsearch_vector.py]: > (cosineSimilarity(params.query_vector, doc['contentVector']) + 1.0) * 0.8: Calculates the cosine similarity between the query vector and the content vector of the document. Cosine similarity measures the cosine of the angle between two vectors and gives a value between -1 and 1. Adding 1.0 ensures that the minimum value is 0.0. The result is then multiplied by 0.8 to give it a weight.

ref: https://medium.com/@bairagiabhishek03/elasticsearch-as-a-vector-store-es-tutorial-5-816f9451ddc1

File: docker/docker-compose.yaml (hunk 3 above)
Comment (miendinh) [reply docker/docker-compose.yaml]: for debugging, I think it's useful for dev.

Comment (bowenliang123) [reply docker/docker-compose.yaml]: -1 for this.

**Comment by JohnJyong** [comment]
=========================== short test summary info ============================
FAILED api/tests/integration_tests/vdb/elasticsearch/test_elasticsearch.py::test_elasticsearch_vector - elasticsearch.exceptions.ConnectionError: ConnectionError(('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))) caused by: ProtocolError(('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer')))
@miendinh  pls help to check

**Comment by bowenliang123** [comment]
Please test on your local development environment first.
Do not keep pushing untested code to run all the CI jobs, which is meaningless and costly.

**Comment by mboo2005** [comment]
@miendinh I tried your branch. Celery run failed because the config is missed. 
error message:
image
Config is not interpreted
image

File: docker/docker-compose.yaml [hunk 5]
```diff
@@ -555,6 +559,26 @@ services:
     ports:
       - "${MYSCALE_PORT:-8123}:${MYSCALE_PORT:-8123}"
 
+  elasticsearch:
+    image: docker.elastic.co/elasticsearch/elasticsearch:7.17.21
```
Comment (bowenliang123) [inline docker/docker-compose.yaml]: Could you use ElasticSearch v8 for testing and running standalone server?

File: web/yarn.lock
Comment (bowenliang123) [inline web/yarn.lock:1]: Please leave the Yarn lockfile unchanged. This PR should not include any changes to frontend code.

File: docker/docker-compose.yaml [hunk 6]
```diff
@@ -555,6 +559,26 @@ services:
...
+      - xpack.security.enabled=false
+    ports:
+      - "9200:9200"
+  kibana:
+    container_name: kb-container
+    image: docker.elastic.co/kibana/kibana:7.17.21
+    environment:
+      - ELASTICSEARCH_HOSTS=http://elasticsearch:9200
+    depends_on:
+      - elasticsearch
```
Comment (bowenliang123) [inline docker/docker-compose.yaml]: And again, Kinbana should not be part of Dify's middleware docker-compose file. Just like Attu, the GUI tool for Milvus  is not shipped with Milvus in the same file.

**Review by bowenliang123 — CHANGES_REQUESTED** [review]
LGTM overall. But several changes are needed for response and actions.

**Comment by miendinh** [comment]
plz check @JohnJyong @crazywoola, thanks.

**Review by crazywoola — DISMISSED** [review]

**Review by crazywoola — APPROVED** [review]
